### ISSUE TITLE
Train model on 4-gpu, the model parameters from different devices are different

### ISSUE THREAD

**Issue by yyht** [uber/horovod#672]
Hi. I think Horovod is a data parallel distributed training framework. Since it is a sync framework, the gradient first averaged via ring-reduce and then broadcast to each device to update parameters of each replica of model. So I think the parameters on different devices should be the same. However, I just store all model parameters via 4-devices and find that the parameters from different devices are different. Did I miss something or mis understand something?

**Comment by tgaddair** [comment]
Hey @yyht, Horovod syncs the _gradients_ between machines, so if the parameters themselves are already out of sync (for example, through random initialization), then the parameters will be different even if the gradients are the same.

One way we ensure that parameters are the same between processes is through the `broadcast` operation.  In TensorFlow, it looks like this:

```
hooks = [hvd.BroadcastGlobalVariablesHook(0)]
with tf.train.MonitoredTrainingSession(hooks=hooks, config=config) as mon_sess:
    ...
```

**Comment by yyht** [comment]
Here is my minimum repro:

init_op = tf.group(tf.global_variables_initializer(), tf.local_variables_initializer())
sess.run(init_op)
print("====succeeded in initializing global and local variables====")
sess.run(hvd.broadcast_global_variables(0))

This my initialization code. I followed the guid that 
"you can simply execute the hvd.broadcast_global_variables op after global variables have been initialized".

Based on this, I first initialization the global variables and then I executed the broadcast_global_variables op.
I was wondering if I did this correctly ?

**Comment by yyht** [comment]
It's werried since I store the parameters after executing "sess.run(hvd.broadcast_global_variables(0))", the parameters from different divices are the same to each other(No training just initialization and boradcasting). But after training, the parameters are different on different devices. 
For data iteration, I just used:
def train_input_fn(input_file, _parse_fn, name_to_features,
		params):
	dataset = tf.data.TFRecordDataset(input_file, buffer_size=params.get("buffer_size", 100))
	print("==hvd size {}, rank {}==".format(hvd.size(), hvd.rank()))
	dataset = dataset.shard(hvd.size(), hvd.rank())
	dataset = dataset.map(lambda x:_parse_fn(x, name_to_features))
	dataset = dataset.shuffle(
							buffer_size=params.get("buffer_size", 1024)+3*params.get("batch_size", 32),
							seed=np.random.randint(0,1e10,1)[0],
							reshuffle_each_iteration=True)
	dataset = dataset.batch(params.get("batch_size", 32))
	dataset = dataset.repeat(params.get("epoch", 100))
	iterator = dataset.make_one_shot_iterator()
	features = iterator.get_next()
	return features

Is it possible that different devices have different iterations(For example, the data size is 249847, for 4-gpu, it will be divided into 4-different subset, and subset with bigger size could have more training iterations leading to this phonemenon?)

**Comment by yyht** [comment]
evaluation after initialization before training:
![image](https://user-images.githubusercontent.com/14133687/49337023-378e8c80-f647-11e8-95c8-c111fbfebb5a.png)
evaluation after one-step model updatation:
![image](https://user-images.githubusercontent.com/14133687/49337025-45dca880-f647-11e8-8f34-7c71ef07c010.png)

the evaluation data iteration code:
        dataset = tf.data.TFRecordDataset(input_file, buffer_size=params.get("buffer_size", 100))
	dataset = dataset.map(lambda x:_parse_fn(x, name_to_features))
	dataset = dataset.batch(params.get("batch_size", 32))
	dataset = dataset.repeat(1)
	iterator = dataset.make_one_shot_iterator()
	features = iterator.get_next()

**Comment by UditGupta10** [comment]
@yyht Are the parameter updates dense like in neural network case or sparse like in word2vec case?

**Comment by yyht** [comment]
parameter updates dense like in neural network case. Actually, I just applied Horovod to BERT model released by google for multi-gpu training.

**Comment by UditGupta10** [comment]
I'll give a perspective here which may or may not be useful. @tgaddair Correct me if am wrong.
The word embeddings are trainable. In one iteration all workers will get different batches to operate on. This will cause different subsets of the embedding matrices to be updated across all the workers. Hence, different parameter values after training.
@yyht Did I miss something?

**Comment by yyht** [comment]
I just print some iteration information about device id and related gloal step. It seems that different devices have different global steps rate that some device are faster than others. Since global step is updated after applying train_op, I think this is not the sync mode? In my opintion, sync mode needs to wait for slowest one to accumulate gradients before updating parameters?
![image](https://user-images.githubusercontent.com/14133687/49337793-f3ee4f80-f653-11e8-8013-37d5b85beefa.png)

**Comment by yyht** [comment]
![image](https://user-images.githubusercontent.com/14133687/49338453-28fea000-f65c-11e8-8dcb-2d76e8ef36d9.png)
Sorry, I made a ### **huge bug** to modify BERT with Horovod that for gradients computation.  I just applied tf.gradients for each woker and apply gradients on each worker that made model on each device different to each other. Thank you for all guys to help with this bugs.

**Comment by tgaddair** [comment]
Glad you got it working @yyht!

To respond to a couple of questions above, if the number of batches is different between workers, then Horovod will fail with a message like "one or more tensors was submitted to be allreduced after a process has finished".  Essentially, the workers with more batches will submit their tensors for allreduce, and will stall waiting for the workers with fewer batches to submit them (which they never do).

@UditGupta10, for sparse tensors like word embeddings (tf.IndexedSlices) we use allgather instead of allreduce to avoid the issue you're describing.  You're definitely right that if you were to do a simple allreduce, you would not get the expected result.

**Comment by yyht** [comment]
for bert model，one node 4 gtx1080Ti vs one gtx1080Ti i got 3.75x speed up，
while on two nodes ，4 gtx1080Ti on each node，i only got similar training speed with one node（i think the bottleneck is on network communication that it needs to do ring-reduce on different nodes and data transportation efficiency depends on your bandwidth i guess）

**Comment by jasstionzyf** [comment]
> ![image](https://user-images.githubusercontent.com/14133687/49338453-28fea000-f65c-11e8-8dcb-2d76e8ef36d9.png)
> Sorry, I made a ### **huge bug** to modify BERT with Horovod that for gradients computation. I just applied tf.gradients for each woker and apply gradients on each worker that made model on each device different to each other. Thank you for all guys to help with this bugs.
i also face same issues, currently i have two gpu, then i run, each gpu's step is not same.   what you means by saying ' I just applied tf.gradients for each woker '  can you paste some code?  currently my code as followers:
` opt = tf.train.AdamOptimizer(learning_rate * hvd.size(), beta1=0.9, beta2=0.999, epsilon=0.1)

    # Op to calculate every variable gradient
    grads = tf.gradients(total_loss, train_variables)
    grads = list(zip(grads, train_variables))

    # Horovod: add Horovod Distributed Optimizer.
    opt = hvd.DistributedOptimizer(opt)
    train_op = opt.apply_gradients(grads_and_vars=grads, global_step=global_step)`
